photon-ml/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala

```diff
@@ -0,0 +1,106 @@
+    *
+    * @return The L2 regularization weight
+    */
+  def l2RegularizationWeight: Double = l2RegWeight
+
+  /**
+    * Protected setter
+    *
+    * @param newRegWeight The new L2 regularization weight
+    */
+  protected[ml] def l2RegularizationWeight_=(newRegWeight: Double): Unit = l2RegWeight = newRegWeight
+
+  abstract override protected[ml] def value(coefficients: Vector[Double]): Double =
+    super.value(coefficients) + l2RegValue(coefficients)
+
+  /**
+    * Compute the L2 regularization value of the given model coefficients
+    *
+    * @param coefficients The model coefficients
+    * @return The L2 regularization value
+    */
+  protected def l2RegValue(coefficients: Vector[Double]): Double =
+    l2RegWeight * coefficients.dot(coefficients) / 2
+}
+
+/**
+  * Trait for a differentiable objective function with L2 regularization.
+  */
+trait L2RegularizationDiff extends DiffFunction with L2Regularization {
+  abstract override protected[ml] def value(coefficients: Vector[Double]): Double =
+    calculate(coefficients)._1
+
+  abstract override protected[ml] def gradient(coefficients: Vector[Double]): Vector[Double] =
+    calculate(coefficients)._2
+
+  abstract override protected[ml] def calculate(coefficients: Vector[Double])
+    : (Double, Vector[Double]) = {
+
+    val (baseValue, baseGradient) = super.calculate(coefficients)
+    (baseValue + l2RegValue(coefficients), baseGradient + l2RegGradient(coefficients))
+  }
+
+  /**
+    * Compute the gradient of the L2 regularization term for the given model coefficients
+    *
+    * @param coefficients The model coefficients
+    * @return The gradient of the L2 regularization term
+    */
+  protected def l2RegGradient(coefficients: Vector[Double]): Vector[Double] =
+    coefficients * l2RegWeight
+}
+
+/**
+  * Trait for a twice differentiable objective function with L2 regularization.
+  */
+trait L2RegularizationTwiceDiff extends TwiceDiffFunction with L2RegularizationDiff {
+  abstract override protected[ml] def hessianVector(
+      coefficients: Vector[Double],
+      multiplyVector: Vector[Double])
+    : Vector[Double] = super.hessianVector(coefficients, multiplyVector) + l2RegHessianVector(multiplyVector)
+
+  abstract override protected[ml] def hessianDiagonal(coefficients: Vector[Double]): Vector[Double] =
+    super.hessianDiagonal(coefficients) + l2RegHessianDiagonal
+
+  /**
+    * Compute the Hessian vector of the L2 regularization term for the given Hessian multiplication vector
+    *
+    * @param multiplyVector The Hessian multiplication vector
+    * @return The Heassian vector of the L2 regularization term
+    */
+  protected def l2RegHessianVector(multiplyVector: Vector[Double]): Vector[Double] =
+    multiplyVector * l2RegWeight
+
+  /**
+    * Compute the Hessian diagonal of the L2 regularization term
+    *
+    * @return The Hessian diagonal of the L2 regularization term
+    */
+  protected def l2RegHessianDiagonal: Double = l2RegWeight
+}
```

Comment:
@joshvfleming 
Moved the old 'withL2Regularization' code into these stackable traits. I'm not sure if this is correct style or not, but it seemed so because unfortunately, as we discussed offline, each descendant of the ObjectiveFunction class needs its own unique behaviour (since each one introduces new methods).

PS - woops, no ending newline